### PR TITLE
Add Rowwise Prune PyTorch op

### PR DESCRIPTION
**Pull request by radkris-git**
Summary:
Add rowwise prune pytorch op.

This operator introduces sparsity to the 'weights' matrix with the help
of the importance indicator 'mask'.

A row is considered important and not pruned if the mask value for that
particular row is 1(True) and not important otherwise.

Test Plan:
buck test caffe2/torch/fb/sparsenn:test -- rowwise_prune
buck test caffe2/test:pruning

Differential Revision: D22849432

### CODE REVIEW

File: aten/src/ATen/native/RowwisePrune.cpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  auto* pruned_2d_tensor_data = pruned_2d_tensor.data_ptr<float>();
+  auto compressed_indices_mapping = at::empty({mask.numel()},
+      compressed_indices_dtype);
+  auto compressed_indices_mapping_data =
+      compressed_indices_mapping.data_ptr<input_t>();
+  auto weights_data = weights.data_ptr<float>();
+  int last_row_kept = 0;
+  for (int i = 0; i < mask.numel(); i++) {
+    if (mask_data[i]) {
+      memcpy(pruned_2d_tensor_data + last_row_kept * num_cols,
```
Comment (raghuramank100) [inline aten/src/ATen/native/RowwisePrune.cpp:33]: Does this work for quantized weights?

File: test/test_pruning_op.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+import hypothesis.strategies as st
+from hypothesis import given
+import numpy as np
+import torch
+from torch.testing._internal.common_utils import TestCase
+
+class PruningOpTest(TestCase):
+
+    def _test_rowwise_prune_op(self, embedding_rows, embedding_dims, indices_type):
```
Comment (raghuramank100) [inline test/test_pruning_op.py:11]: Please add a test to check if this op is scriptable. Since it takes a dtype as an argument, I am not sure if it is

File: aten/src/ATen/native/RowwisePrune.cpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  }
+  int num_cols = weights.size(1);
+  auto pruned_2d_tensor = at::empty({num_non_masked_rows, num_cols},
+      weights.options().dtype(at::kFloat));
+  auto* pruned_2d_tensor_data = pruned_2d_tensor.data_ptr<float>();
+  auto compressed_indices_mapping = at::empty({mask.numel()},
+      compressed_indices_dtype);
+  auto compressed_indices_mapping_data =
+      compressed_indices_mapping.data_ptr<input_t>();
+  auto weights_data = weights.data_ptr<float>();
```
Comment (raghuramank100) [inline aten/src/ATen/native/RowwisePrune.cpp:29]: If this op works only for floating point weights, lets check the dtype of the tensor

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Please see comments

File: test/test_pruning_op.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        embedding_weights = torch.from_numpy((np.random.random_sample((embedding_rows, embedding_dims)) + 1).astype(np.float32))
+        mask = torch.from_numpy(np.random.choice(
```
Comment (supriyar) [inline test/test_pruning_op.py:13]: Should we modify the test to show how the mask can be generated given indicator and threshold values (which can be random values in the test)?

File: aten/src/ATen/native/RowwisePrune.cpp (hunk 1 above)
Comment (radkris-git) [reply aten/src/ATen/native/RowwisePrune.cpp:33]: Good point. Made it generic to work all integral + float types using AT_DISPATCH_ALL_TYPES.

File: test/test_pruning_op.py (hunk 4 above)
Comment (radkris-git) [reply test/test_pruning_op.py:13]: SG. Done

File: test/test_pruning_op.py (hunk 2 above)
Comment (radkris-git) [reply test/test_pruning_op.py:11]: I think this comment doesn't apply as it's just a pytorch c++ op and not a nn.Module.

File: aten/src/ATen/native/RowwisePrune.cpp (hunk 3 above)
Comment (radkris-git) [reply aten/src/ATen/native/RowwisePrune.cpp:29]: Good point. Made it generic with AT_DISPATCH_ALL_TYPES.

**Comment by radkris-git** [comment]
Addressed comments.

File: aten/src/ATen/native/RowwisePrune.cpp [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+  auto mask_data = mask_contig.data_ptr<bool>();
+  for (int i = 0; i < mask.numel(); ++i) {
+    num_non_masked_rows += (((mask_data[i] == true)) ? 1 : 0);
+  }
+  int num_cols = weights.size(1);
+  auto pruned_2d_tensor = at::empty({num_non_masked_rows, num_cols},
+      weights.options());
+  auto compressed_indices_mapping = at::empty({mask.numel()},
+      compressed_indices_dtype);
+  AT_DISPATCH_ALL_TYPES(weights.scalar_type(), "rowwise_prune_helper", [&]() {
```
Comment (supriyar) [inline aten/src/ATen/native/RowwisePrune.cpp:26]: I think this doesn't support at::Half types, can we include that too?

File: test/test_pruning_op.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    # represents the importance of a row.
+    # We mask a row if its indicator value is less than the threshold.
+    def _generate_rowwise_mask(self, embedding_rows):
+        indicator = torch.from_numpy((np.random.random_sample(embedding_rows)).astype(np.float32))
+        threshold = np.random.random_sample()
+        mask = torch.BoolTensor([True if val >= threshold else False for val in indicator])
+        return mask
+
+    def _test_rowwise_prune_op(self, embedding_rows, embedding_dims, indices_type):
+        weight_types = [np.int8, np.int32, np.int16, np.float32, np.double]
```
Comment (supriyar) [inline test/test_pruning_op.py:25]: Can we add this as a hypothesis argument?

Comment (supriyar) [reply test/test_pruning_op.py:25]: Also it would be better to add torch types here rather than numpy types.

Comment (radkris-git) [reply test/test_pruning_op.py:25]: Good points.

**Review by supriyar — APPROVED** [review]
